### ISSUE TITLE
Don't immediately crash when an attribute has an unexpected type

### ISSUE THREAD

**Issue by tmandry** [tmandry/AXSwift#7]
`UIElement.attribute<T>()` takes a generic parameter for its expected return type. For most attributes, the expected type is known, and a failure to convert an existing value to that known type is an error. (Other attributes are allowed to use `attribute<Any>()`.)

I think crashing the program may be a little extreme for an error in this case. There are a couple ways of solving this:

1. Throw a custom error type and let the user handle it if they wish.
2. Another is to remove the casting and generics altogether, and always return `Any` so the user can decide how to handle the cast.
3. Return `nil` if the type is unexpected.

I don't want number 3, because that creates two cases in which the return type could be nil: the attribute doesn't exist, or the attribute exists but is of an unexpected type. Assuming one case or the other could lead to undiscovered bugs in a program.

The main reason I didn't do number 2 is that it makes the API much more awkward to use, in what I expect are the majority of calls. These are cases where there is one expected type, _an unexpected type should trigger some error_, but we don't want to crash the app if the attribute simply isn't there.

Consider:

```swift
// What we have today
let title: String? = element.attribute<String>(.title)
```

vs

```swift
// New signature:
// func attribute(_ attribute: Attribute) -> Any?

// Option 1: Okay, but silently ignores a wrong type!
// What if our program has a bug, or an assumption that isn't always true?
let title: String? = (element.attribute(.title) as? String)

// Option 2: Safer, but uglier and more verbose; requires a temporary
let tmp = element.attribute(.title)
let title: String? = (tmp == nil) ? nil : tmp as! String
```

As it stands now, passing `<Any>` as the type devolves to the second situation, but in that case you don't know the type, and therefore need a more complicated case control structure anyway.

So, I'm leaning towards throwing an error and keeping the API the same. This allows an astute programmer to log the error and devolve functionality gracefully, if possible. But if there's substantial evidence that a lot of use cases _don't_ know their expected type, I might reconsider.

**Comment by acheronfail** [comment]
> So, I'm leaning towards throwing an error and keeping the API the same. This allows an astute programmer to log the error and devolve functionality gracefully, if possible. But if there's substantial evidence that a lot of use cases don't know their expected type, I might reconsider.

FWIW I agree with you on that, I think throwing a custom error is probably best in this scenario. We still get the benefits of how it is, and are protected if the type is unexpected.

**Comment by tmandry** [comment]
Fixed in #27
